### DISCUSSION TITLE
mypy not does not allow positional arguments in RootModel.__init__

### DISCUSSION THREAD

**Discussion by Yasu-umi** [pydantic/pydantic#7418] [Question]
```python
from pydantic import RootModel

class Int(RootModel[int]):
    ...
    
Int(0)
```
This code runs successfully, but mypy produces an "Too many positional arguments" error.
`Int(root=0)` runs successfully & also pass mypy.
Shouldn't it be written as `Int(root=0)` instead of `Int(0)`? (of course I prefer `Int(0)`...)

**Comment by Yasu-umi** [comment]
I try overwrite`__init__`, but it's not working.
```python
from pydantic import RootModel

class IntRootModel(RootModel[int]):
     def __init__(self, root: int) -> None:
         super().__init__(root=root)

IntRootModel(0) # no error

class Int(IntRootModel):
    ...

Int(0) # Too many positional arguments error
```

**Reply by erdnaxeli** [reply]
It is so weird that even when overwriting the constructor the error is still there.

**Reply by Viicos** [reply]
This is because the `__init__` signature is picked up by type checkers only when using the class where it is defined. A similar example with dataclasses (`dataclass_transform` applies the same logic in the example with `IntRootModel`):
```python
from dataclasses import dataclass

@dataclass
class Base:
    a: int

    def __init__(self, custom_field: str) -> None:
        pass

class Sub(Base):
    pass

Base()  # signature hint is `Base(custom_field: str = ...)`
Sub()  # signature hint is `Sub(a: int = ...)`
```

**Reply by edwardpeek-crown-public** [reply]
This doesn't repro for me with mypy 1.7.1 with regular dataclasses, but does with Pydantic models:

```python
from dataclasses import dataclass
from pydantic import BaseModel

@dataclass(kw_only=True)
class DataclassBase:
    member: int

    def __init__(self, member: int) -> None:
        self.member = member

DataclassBase(1)  # no mypy errors

class DataclassSub(DataclassBase):
    pass

DataclassSub(1)  # no mypy errors

class PydanticBase(BaseModel):
    member: int

    def __init__(self, member: int) -> None:
        super().__init__(member=member)

PydanticBase(1)  # no mypy errors

class PydanticSub(PydanticBase):
    pass

PydanticSub(1)  # error: Too many positional arguments for "PydanticSub"  [misc]

```

**Reply by Viicos** [reply]
@edwardpeek-crown According to [PEP 681](https://peps.python.org/pep-0681/#dataclass-semantics):
> As with `dataclass`, method synthesis is skipped if it would overwrite a method that is explicitly declared within the class. Method declarations on base classes do not cause method synthesis to be skipped.

See https://github.com/python/peps/pull/2815 and related issues for more context.

**Reply by edwardpeek-crown-public** [reply]
So the fundamental issue seems to be that with annotated dataclasses you can opt to not re-annotate the derived class and so no new methods will be synthesised to replace the inherited members. However with pydantic it always resynthesises methods unless blocked on each level of derived classes.

**Reply by Viicos** [reply]
The reasoning behind this can be found here: https://github.com/microsoft/pyright/issues/3907#issuecomment-1267710564

If you don't add `@dataclass` on the subclass, it will _not_ be a dataclass, meaning no `__init__` method will be synthesized from annotations, meaning type checkers will use the `__init__` from the parent class.

With `dataclass_transform`, every subclass is considered to have dataclass like behavior, hence the `__init__` method always synthesized.

**Comment by Viicos** [comment]
Probably because `RootModel` inherits from `BaseModel`, where `dataclass_transform` is applied with `kw_only_default`:

https://github.com/pydantic/pydantic/blob/b46010bf4a73d439e3faf98e6e2e16e2faf7e6e2/pydantic/_internal/_model_construction.py#L71-L72

I wonder if there's a way to overcome this issue

**Reply by Yasu-umi** [reply]
Thanks a lot !!!!!!!
I try overriding `dataclass_transform`, but it doesn’t seem to work...
```python
from typing import dataclass_transform

from pydantic import RootModel


class IntRootModel(RootModel[int]):
     def __init__(self, root: int) -> None:
         super().__init__(root=root)

IntRootModel(0) # no error

@dataclass_transform(kw_only_default=False)
class Int(IntRootModel):
    ...

Int(0) # Too many positional arguments error
```

**Reply by Viicos** [reply]
In fact `RootModel` is used with a positional argument in docs, so this is indeed an issue in Pydantic. I'll take a look later in the day.

**Reply by Viicos** [reply]
Since Python 3.7, was dropped, you can now use the following:

```python
IntRootModel = RootModel[int]

# __init__ signature of `IntRootModel`: (root: int = PydanticUndefined, **data: Any) -> None
```

It will not work if you subclass `RootModel` though.

**Comment by artuntun** [comment]
I'm experiencing the exact same issue

**Comment by Viicos** [comment]
A fix was merged in `main` and should land in 2.6, do not hesitate to report any remaining issues once tested on the new release

**Comment by woodruffw** [comment]
Hey all! I'm running `pydantic==2.6.4` with `mypy==1.9.0`, and seeing what I think is a variant of this issue.

Given data models like this:

```py
# example.py

from typing import Union

from pydantic import BaseModel, RootModel


class Bar(BaseModel):
    x: int


class Baz(BaseModel):
    y: str


class Foo(RootModel[Union[Bar, Baz]]):
    root: Union[Bar, Baz]


if __name__ == "__main__":
    bar = Foo(x=1)
    baz = Foo(y="hello")

    print(bar, baz)
```

With `python example.py` I get the expected output:

```
root=Bar(x=1) root=Baz(y='hello')
```

...however, with `mypy example.py`, `mypy` fails to infer that `x` and `y` are valid kwargs for `Foo`:

```
example.py:19: error: Unexpected keyword argument "x" for "Foo"  [call-arg]
example.py:20: error: Unexpected keyword argument "y" for "Foo"  [call-arg]
Found 2 errors in 1 file (checked 1 source file)
```

This goes away if I use the Python 3.8+ annotation syntax, but the reference to a fix above makes me think that the 3.7 syntax should work as well 🙂 

See https://github.com/koxudaxi/datamodel-code-generator/issues/1903 for more context on how this manifests.
